### PR TITLE
run tests when building

### PR DESCRIPTION
**Pull request by jameslamb**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

## Summary

This PR proposes running `{data.table}`'s unit tests as a part of builds here.

This ~introduces 0 new build-time dependencies and~ offers significantly higher coverage of potential build problems than the current testing strategy, which is "check that `library(data.table)` runs without error".

Let's see what the change in build time is when CI runs, but assuming it's on the order of 0-3 minutes, I think it's worth it in exchange for stronger guarantees about correctness of the build here. Especially for a project like `{data.table}`, which is depended on by hundreds of other R packages (see "reverse imports" at https://cran.r-project.org/web/packages/data.table/index.html).

## How I tested this

Locally on my mac (intel chip, macOS 12.2.1, 8GB RAM), ran the following on this branch:

```shell
docker run \
    --rm \
    -v $(pwd):/opt/data.table \
    -w /opt/data.table \
    --entrypoint="" \
    -it condaforge/linux-anvil-cos7-x86_64 \
    bash

conda build .
```

Saw that build succeed, and this output confirming that the tests ran successsfully:

```text
All 8997 tests (last 2163) in tests/tests.Rraw.bz2 completed ok in 25.8s elapsed (28.2s cpu)
+ exit 0

Resource usage statistics from testing r-data.table:
   Process count: 4
   CPU time: Sys=0:00:00.8, User=0:00:27.6
   Memory: 177.3M
   Disk usage: 16B
   Time elapsed: 0:00:28.1
```

For details on that `test.data.table()` function, see https://rdatatable.gitlab.io/data.table/reference/test.data.table.html?q=test.data#ref-examples.

Thanks very much for your time and consideration.

### CODE REVIEW

**Comment by jameslamb** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -43,8 +43,7 @@ requirements:
 
 test:
   commands:
```
Comment (mfansler) [inline recipe/meta.yaml:48]: Might as well enable the optional tests:

```suggestion
  requires:
    - r-bit64
    - r-r.utils
  commands:
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -43,8 +43,7 @@ requirements:
...
-    - $R -e "library('data.table')"           # [not win]
-    - "\"%R%\" -e \"library('data.table')\""  # [win]
+    - Rscript -e "require('data.table'); data.table::test.data.table()"
```
Comment (mfansler) [inline recipe/meta.yaml]: Let's retain an independent loading test and stick to the syntax. Invocation with namespace is sufficient for `test.data.table()`:

```suggestion
    - $R -e "library('data.table')"                   # [not win]
    - $R -e "data.table::test.data.table()"           # [not win]
    - "\"%R%\" -e \"library('data.table')\""          # [win]
    - "\"%R%\" -e \"data.table::test.data.table()\""  # [win]
```

**Review by mfansler — CHANGES_REQUESTED** [review]
Thanks for contributing this! Sounds like an excellent idea.

Some minor suggestions.

File: recipe/meta.yaml (hunk 1 above)
Comment (jameslamb) [reply recipe/meta.yaml:48]: ok works for me! I wasn't sure if maintainers here would be open to adding those extra tests and dependencies.

File: recipe/meta.yaml (hunk 2 above)
Comment (jameslamb) [reply recipe/meta.yaml]: Ok no problem, I wasn't sure how best to set that up.

And I understand now why using variables `$R` / `%R%` that `conda-build` controls are preferable to looking up `Rscript` on `PATH`, sorry about that.

**Comment by mfansler** [comment]
I don't get the error, but for anyone wanting to look into it, here's an extract of the pertinent code to run only the failing test:

**dt-test.2150.025.R**
```r
#!/usr/bin/env Rscript

library(data.table)

dates = as.IDate(c(9610, 19109, 19643, 20385, -1413, 9847, 4116, -11145, -2327, 1760))
times = .POSIXct(tz = 'UTC', c(
  937402277.067304, -626563403.382897, -506636228.039861, -2066740882.02417,
  -2398617863.28256, -1054008563.60793, 1535199547.55902, 2075410085.54399,
  1201364458.72486, 939956943.690777
))
DT = data.table(dates, times)
tmp = tempfile()

fwrite(DT, tmp, dateTimeAs='write.csv')  # as write.csv, writes the UTC times as-is not local because the time column has tzone=="UTC", but without the Z marker

Sys.unsetenv("TZ")
tt = fread(tmp, colClasses=list(POSIXct="times"), tz="")  # from v1.14.0 tz="" needed
data.table:::test(2150.025, attr(tt$times, "tzone"), "")  # as.POSIXct puts "" on the result (testing the write.csv version here with missing tzone)
```

Locally `osx-64` testing fails on R 4.2:

```r
$ conda run -n r42 Rscript dt-test.2150.025.R 
Test 2150.025 ran without errors but failed check that x equals y:
> x = attr(tt$times, "tzone") 
First 0 of 0 (type 'NULL'): 
NULL
> y = "" 
First 1 of 1 (type 'character'): 
[1] ""
target is NULL, current is character

Warning message:
Column 'times' was requested to be 'POSIXct' but fread encountered the following error:
	character string is not in a standard unambiguous format
so the column has been left as type 'character' 
```

but passes on R 4.3

```r
$ conda run -n r43 Rscript dt-test.2150.025.R
Warning message:
In as.POSIXlt.POSIXct(x, tz) :
  datetimes before 1902 may not be accurate: warns once per session
```

 🤷

**Comment by jameslamb** [comment]
Thanks for pulling that out! I'll try to investigate that in the next few days.

I know R 4.3.x contained a bunch of changes related to `POSIX{*}t` classes. From https://cran.r-project.org/doc/manuals/r-release/NEWS.html. I see a few that look promising as root causes, like:

> *as.Date.POSIXct(., tz) now treats several tz values, notably "GMT" as equivalent to "UTC", proposed and improved by Michael Chirico and Joshua Ulrich in [PR#17674](https://bugs.r-project.org/show_bug.cgi?id=17674).*

and 

> *Objects of class "POSIXlt" created in this version of R always have 11 components: component zone is always set, and component gmtoff is set for times in UTC and usually set on the (almost all) platforms which have C-level support, otherwise is NA.*

---

@MichaelChirico if you have time, could you look at https://github.com/conda-forge/r-data.table-feedstock/pull/35#issuecomment-1872365614 and offer any suggestions on where I could start? I'm not asking you to fix this or spend too much time investigating, just hoping to rely on your memory of seeing similar things during `{data.table}` development.

**Comment by MichaelChirico** [comment]
Sorry, I got nothing. It does look like you're barking up the right trees... Can it be minimized to a specific OS and R version? If it is reproducing consistently then it's definitely worth reporting on the data.table tracker, though obviously we'd prefer to have as minimal example as we can...

**Comment by jameslamb** [comment]
No prob, thanks for checking! I'll try to reproduce it outside of conda and report if I find something.

**Comment by jameslamb** [comment]
Here's the content of that temporary CSV file:

<details><summary>data (click me)</summary>

```text
dates,times
1996-04-24,1999-09-15 13:31:17
2022-04-27,1950-02-23 02:36:36
2023-10-13,1953-12-12 03:42:51
2025-10-24,1904-07-05 09:18:37
1966-02-18,1893-12-28 05:15:36
1996-12-17,1936-08-07 19:50:36
1981-04-09,2018-08-25 12:19:07
1939-06-28,2035-10-07 22:48:05
1963-08-19,2008-01-26 16:20:58
1974-10-27,1999-10-15 03:09:03
```

</details>

This test failure is due to exactly 1 of those pieces of data, and I can reproduce the underlying error without `{data.table}`.

```shell
conda create -c conda-forge -y -n r42 'r-base>=4.2,<4.3'
source activate r42
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'

# Error in as.POSIXlt.character(x, tz, ...) : 
#  character string is not in a standard unambiguous format
```

*(that was run on my macbook pro w/ Intel chip, on macOS 12.2.1)*

Some other observations:

1. it doesn't appear that `{data.table}`'s recent release was tested against R 4.2.x on macOS x86_64
    - CRAN checks: https://cran.r-project.org/web/checks/check_results_data.table.html
    - GitHub Actions + Appveyor checks: see the status links on any recent PR, e.g. https://github.com/Rdatatable/data.table/pull/5864
    - on GitLab CI: https://github.com/Rdatatable/data.table/blob/4bc10a649082cca63d8023e79423b1cc85e4a744/.ci/README.md#L9-L19
2. I can't reproduce this using `R-4.2.3.pkg` provided by CRAN (https://cran.r-project.org/bin/macosx/), on the same laptop
    - the only difference I've found so far between those two setups is that the `{data.table}` installation in the conda environment successfully linked against and then used OpenMP, while the one in my native R environment (a precompiled binary from CRAN) is not using OpenMP. I'll investigate that next.

**Comment by MichaelChirico** [comment]
IIUC the MRE youve found doesn't involve data.table code at all? So it looks like a base R bug that we could simply skip in the data.table suite. All the better if you're able to isolate exactly what's going on...

**Comment by jameslamb** [comment]
Yep! Sorry, I juuuuust edited that post by adding "...without `{data.table}`".

Trying to narrow it down more before I report it in `{data.table}`. I think pushing the fix upstream there would be preferable to maintaining a patch here in this conda recipe.

**Comment by jameslamb** [comment]
Comparing the 2 builds of R 4.2.3 (conda vs. native), I don't see any meaningful differences in:

* `Sys.getenv()`
* `options()`
* `R CMD config --all`

I can even see from `R.version` that they're built from the same SVN revision of R.

image

(left is conda, right is native)

I tried running under a very limited environment (only `PATH`), and still found these differences in behavior:

```shell
# --- conda --- #
env -i PATH=/usr/bin:/bin \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# Error in as.POSIXlt.character(x, tz, ...) : 
#  character string is not in a standard unambiguous format

# --- native --- #
env -i PATH=/usr/bin:/bin \
/usr/local/bin/Rscript -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# [1] "1893-12-28 05:15:36 CST"
```

**Comment by jameslamb** [comment]
I think I found the root cause!

## Short Summary

`conda-forge`'s R for macOS was configured with option `--without-internal-tzcode` ([code link](https://github.com/conda-forge/r-base-feedstock/blame/c4525e474a162db58b978f5ed83f051b8ab07c78/recipe/build-r-base.sh#L529)), which tells R to look for timezone names and offsets on the system instead of consulting its own bundled dataset of them.

That differs from the default behavior of CRAN's official R 4.2 binaries for macOS ([docs link](https://cran.r-project.org/doc/manuals/r-devel/R-admin.html#Prerequisites)).

As a result, a call like the following returns `NA` on `conda-forge`'s R 4.2.3 and a correct `POSIXct` object on CRAN's.

```r
strptime(
    unclass("1893-12-28 05:15:36")
    , "%Y-%m-%d %H:%M:%OS"
    , tz = ""
)
```

That difference is why this `{data.table}` test is failing here, and only on the combination of R 4.2 + macOS.

*(I don't understand yet why it's happening on the `x86_64` combination of those and not the `arm64` one... but that can be figured out later)*

This appears to be specific to the case where environment variable `TZ` is unset... passing `TZ=UTC` allows the conversion to succeed.

```shell
env -i PATH=/usr/bin:/bin \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# Error in as.POSIXlt.character(x, tz, ...) : 
#  character string is not in a standard unambiguous format

env -i PATH=/usr/bin:/bin TZ=UTC \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# [1] "1893-12-28 05:15:36 UTC"
```

Passing the string literal `"UTC"` or `"GMT"` to the `tz` keyword argument also works around this.

```shell
env -i PATH=/usr/bin:/bin \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "UTC")'
# [1] "1893-12-28 05:15:36 UTC"

env -i PATH=/usr/bin:/bin \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "GMT")'
# [1] "1893-12-28 05:15:36 GMT"
```

<hr>

## Much More Detailed Summary

<details><summary>investigation and evidence (click me)</summary>

Using the https://github.com/wch/r-source mirror of the R source, I traced backwards through `as.POSIXlt.character()` to figure out where R's behavior deviates on these two builds.

`as.POSIXlt.character()` raises this error if the following test returns `NA` ([code link](https://github.com/wch/r-source/blob/ca87518831d5ead90bc619eb108c5dfdf93cb52d/src/library/base/R/datetime.R#L270-L279))

```r
res <- strptime(
    unclass("1893-12-28 05:15:36")
    , "%Y-%m-%d %H:%M:%OS"
    , tz = ""
)
is.na(res)
```

That returns `NA` on `conda-forge`'s build of R 4.2 but not the native macOSS build from CRAN.

image

*(conda-forge on the left, native on the right)*

`is.na.POSIXlt()` looks like this ([code link](https://github.com/wch/r-source/blob/R-4-2-branch/src/library/base/R/datetime.R#L601-L602)):

```r
is.na.POSIXlt <- function(x)
    is.na(as.POSIXct(x))
```

`as.POSIXct.POSIXlt(x, tz)` runs the following ([code link](https://github.com/wch/r-source/blob/R-4-2-branch/src/library/base/R/datetime.R#L338))

```r
tzone <- attr(x, "tzone")
if(missing(tz) && !is.null(tzone)) tz <- tzone[1L]
y <- .Internal(as.POSIXct(x, tz))
```

As described in https://cran.r-project.org/doc/manuals/r-release/R-ints.html#g_t_002eInternal-vs-_002ePrimitive, `.Internal()` is a call to a C function compiled into R. In this case, `do_asPOSIXct()` ([code link](https://github.com/wch/r-source/blob/e27d62d5b3857ae6a343230d8ebe0ab83ee8332e/src/main/datetime.c#L1132)).

Which notably has this relevant-looking block.

```c
#ifdef USE_INTERNAL_MKTIME
    else R_tzsetwall(); // to get the system timezone recorded
#else
    tzset();
#endif
```

Those appear to be controlled by `configure` options `--with-internal-tzcode / --without-internal-tzcode`.

* https://github.com/wch/r-source/blob/R-4-2-branch/configure.ac#L510-L513
* https://github.com/wch/r-source/blob/R-4-2-branch/configure.ac#L2159-L2169
* https://github.com/wch/r-source/blob/R-4-2-branch/configure.ac#L2524-L2527

As described at https://cran.r-project.org/doc/manuals/r-devel/R-admin.html (my emphasis)

> *Option `--with-internal-tzcode` makes use of R’s own code and copy of the IANA database for managing timezones. This will be preferred where there are issues with the system implementation, usually involving times after 2037 or **before 1916**. An alternative time-zone directory[10](https://cran.r-project.org/doc/manuals/r-devel/R-admin.html#FOOT10) can be used, pointed to by environment variable TZDIR: this should contain files such as Europe/London. On all tested OSes the system timezone was deduced correctly, but if necessary it can be set as the value of environment variable TZ.*

> *configure defaults to `--with-internal-tzcode` on macOS. The native implementation used to be unusable on earlier versions (with a 32-bit time_t and/or timezone tables missing information beyond the 32-bit range). For, e.g., macOS 12.6, option --without-internal-tzcode can be used to override this and R contains sufficient workarounds (for example, the **native code fails to recognize dates with a negative tm_year, that is dates before 1900**) for R to pass its checks. However, there are discrepancies, notably in Europe in the 1900s and 1940s, even though the Olson database contains the correct information.*

`conda-forge`'s build of `r-base` differs from R's default, and passes `--without-internal-tzcode` on macOS ([code link](https://github.com/conda-forge/r-base-feedstock/blob/c4525e474a162db58b978f5ed83f051b8ab07c78/recipe/build-r-base.sh#L529)).

According to `git blame`, it's been like that since the commit that first introduced that recipe 7 years ago ([blame link](https://github.com/conda-forge/r-base-feedstock/blame/c4525e474a162db58b978f5ed83f051b8ab07c78/recipe/build-r-base.sh#L529)).

</details>

## Possible Workarounds

Here's what I can think of. I'm sure others with more `conda` or R internals experience could recommend other options to pursue.

1. `conda-forge`'s `r-base` recipe could be modified to not pass that `--without-internal-tzcode` flag any more
2. `{data.table}` *the library* could be modified to detect R configured this way and adjust for it (e.g. by defaulting to `UTC` for the `tz` keyword argument)
3. `{data.table}`'s tests could be modified (via patches here or in a release commit upstream) to do one of the following:
    - allow this test to fail or skip it entirely on macOS + R4.2
    - or to allow configuration like `tz` to be passed in somehow
    - allow configuration to conditionally omit dates before 1900 and after 2037 
4. this recipe could patch away this particular test
5. the `r-tzdb` package ([conda-forge/r-tzdb-feedstock](https://github.com/conda-forge/r-tzdb-feedstock)) could somehow be used or modified to serve a dataset of timezones that could be referenced by environment variable `TZDIR`, as described in https://cran.r-project.org/doc/manuals/r-devel/R-admin.html#Other-Options
6. running `test.data.table()` here could be deferred, in favor of adding some other testing scripts to the recipe

I'm going to think on this for a bit before recommending one of those.

**Comment by MichaelChirico** [comment]
Thanks! Is it possible to tell via `capabilities()` whether this parameter was set during compilation?

**Comment by MichaelChirico** [comment]
I think (`?capabilities`) that won't work, but I did find this:

> Where R was configured with option `--with-internal-tzcode` (the default on Windows: recommended on Solaris), the database at `file.path(R.home("share"), "zoneinfo")` is used by default: file `VERSION` in that directory states the version.

I wonder if that directory is maybe missing under `--without-internal-tzcode`? Which could be something to check...

**Comment by MichaelChirico** [comment]
One more comment -- it does look like a `conda` bug that this compilation setting makes results vary across systems/R versions, as IIUC one goal of conda is to allow consistency without user having to worry about such details.

Perhaps `conda` needs to supply `TZDIR=` to point to a specific conda-internal timezone info database? Otherwise using the R-provided database (i.e. restoring `--with-internal-tzcode`) seems like the proper thing to do.

**Comment by jameslamb** [comment]
> I wonder if that directory is maybe missing under `--without-internal-tzcode`?

Looks like it is missing.

In a `conda` environment with `conda-forge`'s R 4.2, I see the following.

```shell
env -i PATH=/usr/bin:/bin LANG=en_US.UTF-8 \
"$CONDA_PREFIX/bin/Rscript" -e 'print(file.path(R.home("share"), "zoneinfo"))'
# "/Users/jlamb/mambaforge/envs/r42/lib/R/share/zoneinfo"

ls /Users/jlamb/mambaforge/envs/r42/lib/R/share/zoneinfo
# ls: /Users/jlamb/mambaforge/envs/r42/lib/R/share/zoneinfo: No such file or directory

find $CONDA_PREFIX -name 'zoneinfo' -type d
# (empty)
```

> Perhaps conda needs to supply `TZDIR`= to point to a specific conda-internal timezone info database?

Yeah, to work with these pre-1900 dates, it would be better to point R at some other suitable set of timezone information via `TZDIR`.

For example, the IANA timezone database is packaged for `conda-forge` at https://github.com/conda-forge/tzdata-feedstock.

I tried installing it and pointing R at it via `TZIDR`. That unfortunately did not work.

```shell
conda install --yes -n r42 tzcode tzdata

find $CONDA_PREFIX -name 'zoneinfo' -type d
# /Users/jlamb/mambaforge/envs/r42/share/zoneinfo

env -i PATH=/usr/bin:/bin \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# Error in as.POSIXlt.character(x, tz, ...) : 
#  character string is not in a standard unambiguous format

env -i PATH=/usr/bin:/bin LANG=en_US.UTF-8 TZDIR="$CONDA_PREFIX/share/zoneinfo" \
"$CONDA_PREFIX/bin/Rscript" -e 'as.POSIXct("1893-12-28 05:15:36", tz = "")'
# Error in as.POSIXlt.character(x, tz, ...) : 
#  character string is not in a standard unambiguous format
```

I do see a timezone file for my timezone there.

```shell
env -i PATH=/usr/bin:/bin LANG=en_US.UTF-8 TZDIR="$CONDA_PREFIX/share/zoneinfo" \
"$CONDA_PREFIX/bin/Rscript" -e 'print(Sys.timezone())'
# [1] "America/Chicago"

ls $CONDA_PREFIX/share/zoneinfo/America
```

Screen Shot 2024-01-02 at 8 48 46 PM

So maybe there is more going on here?

**Comment by jameslamb** [comment]
> it does look like a conda bug that this compilation setting makes results vary across systems/R versions, as IIUC one goal of conda is to allow consistency without user having to worry about such details.

I agree with this... I think it would be preferable for `conda`'s `r-base` for macOS to distribute R's timezone database.

But I also don't think that should block improving `conda-forge`'s `data.table` testing... the patch you put up in https://github.com/Rdatatable/data.table/pull/5867 shoulld be enough for that.

My proposed sequence of events:

1. I'll put up an issue suggesting that change to `conda-forge`s R shortly
2. you merge https://github.com/Rdatatable/data.table/pull/5867 if you get agreement from other `{data.table}` maintainers there
3. whenever a `{data.table}` release with that `test.data.table()` patch gets to CRAN, I update this PR to target that version
4. hopefully everything passes and we merge this PR

**Comment by jameslamb** [comment]
I've opened https://github.com/conda-forge/r-base-feedstock/issues/290.

Thanks for all your help narrowing this down and finding potential ways forward @MichaelChirico !

**Review by mfansler — APPROVED** [review]
Looks good to go! Thanks everyone for helping with this! ❤️

**Comment by mfansler** [comment]
@jameslamb feel free to tag me in any future R testing PRs - always like to have more confidence in our builds!

**Comment by jameslamb** [comment]
Will do, thank you!

**Pull request merged**